### PR TITLE
[PHI Kernels] add pad op for xpu

### PR DESCRIPTION
**Pull request by lj970926**
### PR types

Others

### PR changes

OPs

#### 为XPU添加pad_kernel及对应的反向算子
[nn.functional.pad](https://github.com/PaddlePaddle/Paddle/blob/f023d42f36ee0e7bfb1c8162987a5b40dd97cae1/python/paddle/nn/functional/common.py#L1449)在```mode="constant"```和```len(pad) == dim * 2```的情况下会调用pad op。

该Op在[CPU](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/cpu/pad_kernel.cc)和[GPU](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/cpu/pad_kernel.cc)中复用了[paddle/phi/kernels/func/padding.h](https://github.com/PaddlePaddle/Paddle/blob/f023d42f36ee0e7bfb1c8162987a5b40dd97cae1/paddle/phi/kernels/funcs/padding.h#LL34C23-L34C23)中基于EigenTensor的实现，EigenTensor不支持XPU，需要调用xdnn的pad api实现。

单测部分的代码参考了https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/test_pad_op.py

### CODE REVIEW

File: paddle/phi/kernels/xpu/pad_grad_kernel.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  XPUType value = static_cast<XPUType>(pad_value.to<T>());
+  int r = xpu::pad<XPUType>(dev_ctx.x_context(),
+                            reinterpret_cast<const XPUType*>(d_out.data<T>()),
+                            reinterpret_cast<XPUType*>(d_x->data<T>()),
+                            out_shape,
+                            pad_left,
+                            pad_right,
+                            value);
+  PADDLE_ENFORCE_XDNN_SUCCESS(r, "pad");
```
Comment (RuohengMa) [inline paddle/phi/kernels/xpu/pad_grad_kernel.cc:45]: 应该是pad_grad~

File: paddle/phi/kernels/xpu/pad_grad_kernel.cc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/pad_grad_kernel.h"
+
+#include "paddle/phi/backends/xpu/enforce_xpu.h"
+#include "paddle/phi/backends/xpu/xpu_context.h"
```
Comment (ykkk2333) [inline paddle/phi/kernels/xpu/pad_grad_kernel.cc]: 这个头文件可以不用

**Review by ykkk2333 — DISMISSED** [review]
LGTM

File: test/xpu/test_pad_op_xpu.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+                def test_Variable():
+                    paddle.nn.functional.pad(x=input_data, pad=[1, 1, 1, 1])
+
+                self.assertRaises(TypeError, test_Variable)
+
+                data = paddle.static.data(
+                    name='data', shape=[4], dtype='float16'
+                )
+                paddle.nn.functional.pad(x=data, pad=[0, 1])
```
Comment (RuohengMa) [inline test/xpu/test_pad_op_xpu.py:116]: 对于xpu算子允许范围以外的数据类型，我不确定单测里需不需要测，需要看看@ykkk2333怎么说

File: test/xpu/test_pad_op_xpu.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                    feat = fc(x)  # [2,3,10]
+
+                    out = self.call_func(feat)
+
+                    sgd = paddle.optimizer.SGD()
+                    sgd.minimize(paddle.mean(out))
+                    self.assertTrue(self.var_prefix() in str(main_prog))
+                    exe = paddle.static.Executor(paddle.XPUPlace(0))
+                    exe.run(starup_prog)
+                    res = exe.run(fetch_list=[feat, out])
```
Comment (RuohengMa) [inline test/xpu/test_pad_op_xpu.py:142]: 这里为什么要实际跑一次训练？上面TestCase1-3测的和这里有什么不同？

File: paddle/phi/kernels/xpu/pad_grad_kernel.cc (hunk 2 above)
Comment (lj970926) [reply paddle/phi/kernels/xpu/pad_grad_kernel.cc]: Done

File: paddle/phi/kernels/xpu/pad_grad_kernel.cc (hunk 1 above)
Comment (lj970926) [reply paddle/phi/kernels/xpu/pad_grad_kernel.cc:45]: 请问这里不应该是和实际调用的XDNN API保持一致吗，pad_grad_kernel里面实际调用的也是xpu::pad

File: test/xpu/test_pad_op_xpu.py (hunk 3 above)
Comment (lj970926) [reply test/xpu/test_pad_op_xpu.py:116]: XPU实际是支持FP16的，已经加上了

**Review by RuohengMa — DISMISSED** [review]
LGTM

File: test/xpu/test_pad_op_xpu.py (hunk 3 above)
Comment (RuohengMa) [reply test/xpu/test_pad_op_xpu.py:116]: ok

File: paddle/phi/kernels/xpu/pad_grad_kernel.cc (hunk 1 above)
Comment (RuohengMa) [reply paddle/phi/kernels/xpu/pad_grad_kernel.cc:45]: 好的~

File: test/xpu/test_pad_op_xpu.py (hunk 4 above)
Comment (lj970926) [reply test/xpu/test_pad_op_xpu.py:142]: 这个是我看CPU和GPU的单测里面有就加上了，看内容似乎是和框架里面的其他部分一起做的一个小的集成测试？

**Review by RuohengMa — APPROVED** [review]
LGTM

**Review by houj04 — APPROVED** [review]
